src/main/kotlin/com/geekbeast/rhizome/jobs/AbstractDistributedJob.kt

```diff
@@ -0,0 +1,260 @@
+         */
+
+        status = jobs.executeOnKey(id!!) { it.value.status }
+
+        if (resumable && status == JobStatus.RUNNING) {
+            //Case 1
+            val v = jobs.executeOnKey(id!!) { it.value }!!
+            this.state = state
+            this.status = v.status
+            this.taskId = v.taskId
+            this.hasWorkRemaining = v.hasWorkRemaining
+            this.progress = progress
+        } else if (status == JobStatus.PENDING) {
+            //If status is pending then job is new and we just need to mark as running.
+            status = JobStatus.RUNNING
+        }
+    }
+
+    private fun processBatches() {
+        while (!interrupted() && hasWorkRemaining && status == JobStatus.RUNNING) {
+            processNextBatch()
+            if (!hasWorkRemaining) {
+                status = JobStatus.FINISHED
+            }
+            publishJobState()
+        }
+    }
+
+    final override fun call(): R? {
+        //Wait until task id has been assigned and can be retrieved.
+        initializeTaskId()
+
+        /**
+         * If the job status is paused we poll for a job status update every five seconds, otherwise we process
+         * that batches, which publishes it's state on every loop. If we get interrupted, we abort the processing
+         * and publish the aborted job state.
+         */
+        try {
+            do {
+                if (status == JobStatus.PAUSED) {
+                    Thread.sleep(PAUSE_POLLING_MILLIS)
+                }
+                initializeTask()
+                processBatches()
+            } while (resumable && status == JobStatus.PAUSED)
+        } catch (ex: InterruptedException) {
+            logger.warn("Distributed job $taskId was interrupted.", ex)
+            abort()
+        } finally {
+            publishJobState()
+        }
+
+        return if (hasWorkRemaining && status == JobStatus.CANCELED) {
+            logger.warn("Distributed job $id with task id $taskId was interrupted or canceled.")
+            null
+        } else if (hasWorkRemaining && resumable && status == JobStatus.PAUSED) {
+            logger.info("Distributed job $id with task id $taskId was pasused")
+            null
+        } else if (!hasWorkRemaining && status == JobStatus.FINISHED) {
+            result
+        } else {
+            throw IllegalStateException("Distributed job $id is an illegal state.")
+        }
+    }
+
+    private fun initializeTaskId() {
+        if (taskId != null) return
+
+        require(id != null) { "Cannot initialize task when id has not been initialized." }
+
+        while( taskId == null ) {
+            val maybeTaskId = jobs.executeOnKey(id!!) { it.value.taskId } //TODO: Make offloadable
+            if (maybeTaskId == null) Thread.sleep(INITIALIZE_TASK_ID_POLLING_MILLIS) else initTaskId(maybeTaskId)
+        }
+    }
+
+    protected fun publishJobState() {
+        //Do not replace with indexing operator
```

Comment:
indexing operator?

Reply:
jobs[id] = this, intellij recommends it, but [] invokes put, which will cause reading out the existing value unlike set which does not.